### PR TITLE
Add --check flag to tree-sitter highlight.

### PR DESCRIPTION
**Pull request by patrickt**
Recently I've been pulling a lot of grammars into GitHub's highlighting backend,
replacing legacy language support with tree-sitter highlighting queries.
Our backend systems have a standard set of highlight captures we expect, very
similar to the standard tagging captures we expect. Though end-user applications
are free to choose whatever tagging nomenclature they want, I think it's nice to
include a checking stage that will help us ensure that we know whether a capture
might be recognized or not. It will also help us figure out where we need to
expand our standard set of captures (see #1539).

### CODE REVIEW

File: highlight/src/lib.rs [hunk 1]
```diff
@@ -14,6 +16,41 @@ const CANCELLATION_CHECK_INTERVAL: usize = 100;
...
+    static ref STANDARD_CAPTURE_NAMES: HashSet<&'static str> = vec![
+        "attribute",
+        "carriage-return",
+        "comment",
+        "constant",
+        "constant.builtin",
+        "constructor",
+        "constructor.builtin",
+        "embedded",
+        "escape",
```
Comment (amaanq) [inline highlight/src/lib.rs:29]: Hmm, why not string.escape? would be a bit more clear and allows escape sequences to fallback to string if inside one

File: highlight/src/lib.rs [hunk 2]
```diff
@@ -14,6 +16,41 @@ const CANCELLATION_CHECK_INTERVAL: usize = 100;
 const BUFFER_HTML_RESERVE_CAPACITY: usize = 10 * 1024;
 const BUFFER_LINES_RESERVE_CAPACITY: usize = 1000;
 
+lazy_static! {
+    static ref STANDARD_CAPTURE_NAMES: HashSet<&'static str> = vec![
+        "attribute",
+        "carriage-return",
+        "comment",
```
Comment (amaanq) [inline highlight/src/lib.rs:23]: `@comment.documentation` and `@string.documentation` would be nice to add too

**Comment by clason** [comment]
Hi @patrickt  👋

How does this tie into coordinating queries across consumers? I am in favor of any upstream capability that makes this easier, but am a bit wary about baking in one specific ecosystem into a prima facie agnostic infrastructure.

**Comment by clason** [comment]
For the record here are the captures used by 
* [Neovim](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights)
* [Helix](https://docs.helix-editor.com/master/themes.html#syntax-highlighting)

It would be great if these were to be given equal weight; e.g., only the lowest common denominator would be enforced.

(Neovim is discussing [closer alignment with Helix/upstream captures](https://github.com/nvim-treesitter/nvim-treesitter/issues/4799), but full coincidence will likely never been possible -- nor reasonable. Nevertheless, such an upstream check can accelerate this process and prevent regressions on agreed alignment, so it is not unwelcome if done properly.)

**Comment by patrickt** [comment]
@clason Great questions! I think it’s an excellent idea to adopt the common subset, or something close thereto, of capture names used by GitHub (which is the contents of this patch’s original list), Helix, Neovim, and Emacs. I also agree that perfect alignment between all projects and across all grammrs is neither possible nor desirable: this is why I put this command under the `highlight` and not the `test` subcommand, as I wanted the result to be an informational, not prescriptive, datum. I’m going to take a look this week and see what additions and subtractions are going to be the most useful. I’m glad you think that this is going to be useful to help prevent highlighting regressions, and I also hope that this is going to help authors of new grammars figure out what kind of entities are most useful to highlight.

**Comment by patrickt** [comment]
Proposed additions:
* `string.escape`
* `string.regexp`
* `type.enum`
* `string.special.symbol`
* `comment.documentation`
* `variable.member`
* `label`
* `error`
* the `markup` hierarchy, as it looks like the Neovim folks are moving towards `markup` and away from `text.

**Comment by amaanq** [comment]
A few more to consider - though not all of them are entirely necessary, I think some of them would be nice additions

- `type.qualifier`
- `keyword.coroutine`
- `keyword.function`
- `keyword.operator`
- `keyword.return`
- `repeat`
- `conditional`
- `include`
- `boolean`
- `float`
- `character`
- `preproc`
- `storageclass`
- `field/property`
- `string/comment.documentation`

**Comment by clason** [comment]
Another point where alignment is relevant is predicates and directives (knowing that the latter are not officially part of tree-sitter) -- here, too, a "common core" would be useful to agree on.

(Also pinging @the-mikedavis for Helix.)

**Comment by clason** [comment]
Oh, and yet another thing (which you are but readers may or may not be aware of): The query precedence in tree-sitter CLI and Neovim is different:
1. tree-sitter CLI has "first match wins" (since a match terminates query processing);
2 Neovim has "last match wins" (since all queries are always processed -- and applied -- in order, and later queries "paint over" previous ones).

([Zed seems to follow option 2 as well](https://github.com/nvim-treesitter/nvim-treesitter/issues/3944#issuecomment-1458782497).)

So it would be immensely helpful if the behavior of `tree-sitter highlight` and `tree-sitter test` could be controlled by a flag like `--no-skip` to replicate Neovim's (and Zed's) behavior. 

(Not in this PR, of course; but I wanted to bring this up since it fits into the general "alignment" strategy.)

File: highlight/src/lib.rs [hunk 3]
```diff
@@ -321,6 +381,14 @@ impl HighlightConfiguration {
                 best_index.map(Highlight)
             }));
     }
+
+    pub fn nonconformant_capture_names(&self) -> Vec<&String> {
+        return self
+            .names()
+            .iter()
+            .filter(|&n| !STANDARD_CAPTURE_NAMES.contains(n.as_str()))
```
Comment (the-mikedavis) [inline highlight/src/lib.rs:389]: What do you think about also filtering out `n.starts_with('_')`? I've seen this convention used in a few languages with more complex queries to be able to use a value in a `#eq?`/`#match?` predicate without attaching a "real" highlight, like https://github.com/elixir-lang/tree-sitter-elixir/blob/7be390548a870ca9cb1bd7f59ac92457bbec7bf5/queries/highlights.scm#L93-L97

(Across all Helix queries: https://github.com/search?q=repo%3Ahelix-editor%2Fhelix+%2F%40_%5Cw%2B%2F+path%3Aruntime%2Fqueries&type=code)

Comment (clason) [reply highlight/src/lib.rs:389]: Yes, captures starting with `_` are treated as "private" by nvim-treesitter's Capture Police as well.

Comment (patrickt) [reply highlight/src/lib.rs]: A great idea. Incorporated.

**Comment by clason** [comment]
Another thought: would it be possible/desirable to allow specifying additional allowed captures in a config file (to be passed, e.g., via a flag)? That way Helix and nvim-treesitter could include their own set into the check and use it for their CI. (Nvim-treesitter uses a [script that scrapes documentation](https://github.com/nvim-treesitter/nvim-treesitter/blob/1ef286e5b0cfd17f56586a8445cd83d61647f851/scripts/check-queries.lua#L12-L36) for that.)

**Comment by patrickt** [comment]
@clason I think that would be an excellent idea, but is probably better suited for a follow-up patch.

**Comment by clason** [comment]
Of course, just wanted to note this down lest I forget. I'll open an issue with the follow-up ideas once this is merged.

**Review by amaanq — APPROVED** [review]

**Comment by amaanq** [comment]
CI will work upon a rebase @patrickt

**Review by ahlinc — APPROVED** [review]

**Pull request merged**